adr/0025-required-optional.md

```diff
@@ -0,0 +1,127 @@
+# 25. Components can be required by default + introduction of feature flags
+
+Date: 2024-04-29
+
+## Status
+
+Accepted
+
+## Context
+
+> Feature request: <https://github.com/defenseunicorns/zarf/issues/2059>
+
+Currently, all Zarf components default to being optional due to the `required` key being _optional_ in the YAML. This leads to package authors needing to ensure that they annotate this key for each component, and since nothing in the current validations prompts them about this they may be confused about the "all things are optional" default state.
+
+When Zarf was first created, we didn't really know how it would evolve and this key was introduced in those very early days. At this point it would be better to require all components by default--especially with the introduction of composability and the OCI skeleton work, there is plenty of flexibility in the API to compose bespoke packages assembled from other packages.
+
+A few ways to handle this:
+
+1. Simply force the `required` key to be a non-optional, so that package authors would be forced to specify it for each component, thereby removing any ambiguity--but also force one more key for every single component ever created 🫠
+
+2. Deprecate `required` and introduce an optional `optional` key, which would default to _false_.
+
+3. Do something more significant like combine various condition-based things such as `only`, `optional` (instead of `required`), or `default`.
+
+4. Introduce `feature` flags to allow for certain schema behavior to be configurable to the user.
+
+## Decision
+
+Option 4. Introduce `.metadata.features` flags to `zarf.yaml` to allow for certain schema behavior to be configurable to the user.
+
+The `features` key will be added to the `metadata` section of the package schema. This key will be an array of strings, where each string is the name of a beta feature that can be enabled. To enable a feature, the user will need to add the name of the feature to this array and edit the package schema accordingly.
```

Comment:
I believe the `.metadata.features` field might not be necessary for this purpose. Have we considered keeping the `required` key optional in the schema and defaulting to `required: true` for all components unless explicitly set to `required: false`? This would achieve the same behavior as the `default-required` flag but without needing the feature flag.